File: src/skills/buffett/scripts/calc_bollinger.py
"""用后复权收盘价计算布林带（供 buffett Step 4）。

默认参数与 SKILL.md 一致：
  - 日线/周线：20 期，2 倍标准差
  - 月线：24 期，2 倍标准差

输入（任选其一）：
  1) 东财 `opencli eastmoney kline ... --adjust backward -f json`（字段 date/close，或包在数组/带 rows 的对象里）
  2) fetch_hfq_daily.py 产出的 JSON（bars[].trade_date / close）
  3) stdin 同结构 JSON

用法:
  opencli eastmoney kline 600900 --period day --adjust backward --limit 520 -f json > /tmp/k.json
  python3 calc_bollinger.py /tmp/k.json
  python3 calc_bollinger.py /tmp/hfq.json --period D --window 20 --nbdev 2
"""

from __future__ import annotations

import argparse
import json
import math
import re
import sys
from typing import Any


DEFAULTS = {
    "D": {"window": 20, "nbdev": 2.0},
    "W": {"window": 20, "nbdev": 2.0},
    "M": {"window": 24, "nbdev": 2.0},
}


def load_payload(path: str | None) -> dict[str, Any] | list[Any]:
    if path:
        with open(path, encoding="utf-8") as fh:
            return json.load(fh)
    return json.load(sys.stdin)


def normalize_date(raw: Any) -> str | None:
    if raw is None:
        return None
    s = str(raw).strip()
    if not s:
        return None
    digits = re.sub(r"\D", "", s)
    if len(digits) >= 8:
        return digits[:8]
    return None


def to_bars(payload: dict[str, Any] | list[Any]) -> list[dict[str, Any]]:
    if isinstance(payload, list):
        return payload
    for key in ("bars", "data", "rows", "items", "kline", "klines"):
        value = payload.get(key)
        if isinstance(value, list):
            return value
    raise ValueError(
        "输入 JSON 须为 K 线数组，或含 bars/data/rows[]（东财 kline / Tushare fetch_hfq）"
    )


def resample_closes(
    bars: list[dict[str, Any]], period: str
) -> list[tuple[str, float]]:
    """period: D|W|M。周/月取该周期最后一个交易日的后复权收盘价。"""
    daily: list[tuple[str, float]] = []
    for row in bars:
        date = normalize_date(row.get("trade_date") or row.get("date") or row.get("day"))
        close = row.get("close")
        if not date or close is None:
            continue
        try:
            c = float(close)
        except (TypeError, ValueError):
            continue
        if math.isnan(c):
            continue
        daily.append((date, c))
    daily.sort(key=lambda x: x[0])

    if period == "D":
        return daily

    buckets: dict[str, tuple[str, float]] = {}
    for date, close in daily:
        y, m, d = int(date[0:4]), int(date[4:6]), int(date[6:8])
        if period == "M":
            key = f"{y:04d}{m:02d}"
        else:  # W — ISO 周
            import datetime as dt

            key = dt.date(y, m, d).strftime("%G-W%V")
        # 同桶保留较晚交易日
        prev = buckets.get(key)
        if prev is None or date >= prev[0]:
            buckets[key] = (date, close)

    return [buckets[k] for k in sorted(buckets)]


def bollinger(
    series: list[tuple[str, float]], window: int, nbdev: float
) -> dict[str, Any]:
    if len(series) < window:
        return {
            "ok": False,
            "error": f"有效样本 {len(series)} < window={window}，无法计算布林带",
            "count": len(series),
        }

    closes = [c for _, c in series]
    # 用尾部 window 根算当前轨（与 skill「当前位置」一致）
    window_closes = closes[-window:]
    mean = sum(window_closes) / window
    var = sum((x - mean) ** 2 for x in window_closes) / window  # 总体标准差（常用交易口径）
    std = math.sqrt(var)
    upper = mean + nbdev * std
    lower = mean - nbdev * std
    last_date, last_close = series[-1]
    bandwidth = (upper - lower) / mean if mean else None

    return {
        "ok": True,
        "as_of": last_date,
        "close": round(last_close, 4),
        "mid": round(mean, 4),
        "upper": round(upper, 4),
        "lower": round(lower, 4),
        "bandwidth": None if bandwidth is None else round(bandwidth, 6),
        "bandwidth_pct": None if bandwidth is None else round(bandwidth * 100, 2),
        "window": window,
        "nbdev": nbdev,
        "sample_count": len(series),
    }


def main() -> int:
    parser = argparse.ArgumentParser(description="后复权收盘价布林带（buffett Step 4）")
    parser.add_argument(
        "input",
        nargs="?",
        help="东财 kline / fetch_hfq_daily JSON 路径；省略则读 stdin",
    )
    parser.add_argument(
        "--period",
        choices=["D", "W", "M", "all"],
        default="all",
        help="D 日 / W 周 / M 月 / all 三者（默认 all）",
    )
    parser.add_argument("--window", type=int, help="覆盖默认窗口")
    parser.add_argument("--nbdev", type=float, help="覆盖默认标准差倍数")
    args = parser.parse_args()

    try:
        payload = load_payload(args.input)
        bars = to_bars(payload)
    except Exception as exc:  # noqa: BLE001
        print(json.dumps({"error": str(exc)}, ensure_ascii=False), file=sys.stderr)
        return 1

    meta = payload if isinstance(payload, dict) else {}
    periods = ["D", "W", "M"] if args.period == "all" else [args.period]
    result: dict[str, Any] = {
        "ts_code": meta.get("ts_code") or meta.get("symbol"),
        "adj": meta.get("adj") or "backward/hfq",
        "bands": {},
    }

    for p in periods:
        defaults = DEFAULTS[p]
        window = args.window or defaults["window"]
        nbdev = args.nbdev if args.nbdev is not None else defaults["nbdev"]
        series = resample_closes(bars, p)
        result["bands"][p] = {
            "period": p,
            **bollinger(series, window, nbdev),
        }

    print(json.dumps(result, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/skills/buffett/scripts/calc_bollinger.js
#!/usr/bin/env node
/**
 * 用后复权收盘价计算布林带（供 buffett Step 3）。
 *
 * 默认参数与 SKILL.md 一致：
 *   - 日线/周线：20 期，2 倍标准差
 *   - 月线：24 期，2 倍标准差
 *
 * 用法:
 *   node calc_bollinger.js tmp/k.json
 *   node fetch_kline_hfq.js 600900 --browser-only -o tmp/k.json && node calc_bollinger.js tmp/k.json
 *   node calc_bollinger.js tmp/k.json --period D --window 20 --nbdev 2
 */

import fs from "node:fs";
import { parseArgs } from "./opencli_json.js";

const DEFAULTS = {
  D: { window: 20, nbdev: 2.0 },
  W: { window: 20, nbdev: 2.0 },
  M: { window: 24, nbdev: 2.0 },
};

function loadPayload(path) {
  const text = path ? fs.readFileSync(path, "utf8") : fs.readFileSync(0, "utf8");
  return JSON.parse(text);
}

function normalizeDate(raw) {
  if (raw == null) return null;
  const s = String(raw).trim();
  if (!s) return null;
  const digits = s.replace(/\D/g, "");
  return digits.length >= 8 ? digits.slice(0, 8) : null;
}

function toBars(payload) {
  if (Array.isArray(payload)) return payload;
  for (const key of ["bars", "data", "rows", "items", "kline", "klines"]) {
    if (Array.isArray(payload?.[key])) return payload[key];
  }
  throw new Error("输入 JSON 须为 K 线数组，或含 bars/data/rows[]（东财 kline --adjust backward）");
}

/** ISO 周键 YYYY-Www */
function isoWeekKey(y, m, d) {
  const date = new Date(Date.UTC(y, m - 1, d));
  // Thursday in current week decides the year
  const dayNum = date.getUTCDay() || 7;
  date.setUTCDate(date.getUTCDate() + 4 - dayNum);
  const yearStart = new Date(Date.UTC(date.getUTCFullYear(), 0, 1));
  const weekNo = Math.ceil(((date - yearStart) / 86400000 + 1) / 7);
  const isoYear = date.getUTCFullYear();
  return `${isoYear}-W${String(weekNo).padStart(2, "0")}`;
}

/** period: D|W|M。周/月取该周期最后一个交易日的后复权收盘价。 */
function resampleCloses(bars, period) {
  const daily = [];
  for (const row of bars) {
    const date = normalizeDate(row.trade_date || row.date || row.day);
    const close = row.close;
    if (!date || close == null) continue;
    const c = Number(close);
    if (!Number.isFinite(c)) continue;
    daily.push([date, c]);
  }
  daily.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));

  if (period === "D") return daily;

  const buckets = new Map();
  for (const [date, close] of daily) {
    const y = Number(date.slice(0, 4));
    const m = Number(date.slice(4, 6));
    const d = Number(date.slice(6, 8));
    const key = period === "M" ? `${String(y).padStart(4, "0")}${String(m).padStart(2, "0")}` : isoWeekKey(y, m, d);
    const prev = buckets.get(key);
    if (!prev || date >= prev[0]) buckets.set(key, [date, close]);
  }
  return [...buckets.entries()].sort((a, b) => (a[0] < b[0] ? -1 : 1)).map(([, v]) => v);
}

function bollinger(series, window, nbdev) {
  if (series.length < window) {
    return {
      ok: false,
      error: `有效样本 ${series.length} < window=${window}，无法计算布林带`,
      count: series.length,
    };
  }
  const closes = series.map(([, c]) => c);
  const windowCloses = closes.slice(-window);
  const mean = windowCloses.reduce((a, b) => a + b, 0) / window;
  const variance = windowCloses.reduce((a, x) => a + (x - mean) ** 2, 0) / window;
  const std = Math.sqrt(variance);
  const upper = mean + nbdev * std;
  const lower = mean - nbdev * std;
  const [lastDate, lastClose] = series[series.length - 1];
  const bandwidth = mean ? (upper - lower) / mean : null;

  return {
    ok: true,
    as_of: lastDate,
    close: round(lastClose, 4),
    mid: round(mean, 4),
    upper: round(upper, 4),
    lower: round(lower, 4),
    bandwidth: bandwidth == null ? null : round(bandwidth, 6),
    bandwidth_pct: bandwidth == null ? null : round(bandwidth * 100, 2),
    window,
    nbdev,
    sample_count: series.length,
  };
}

function round(n, digits) {
  const p = 10 ** digits;
  return Math.round(n * p) / p;
}

function main() {
  const args = parseArgs(process.argv.slice(2), {
    positional: ["input"],
    defaults: { period: "all" },
  });

  let payload;
  let bars;
  try {
    payload = loadPayload(args.input);
    bars = toBars(payload);
  } catch (exc) {
    console.error(JSON.stringify({ error: String(exc.message || exc) }));
    return 1;
  }

  const meta = payload && typeof payload === "object" && !Array.isArray(payload) ? payload : {};
  const periods = args.period === "all" ? ["D", "W", "M"] : [args.period];
  if (!["D", "W", "M", "all"].includes(args.period)) {
    console.error(JSON.stringify({ error: `--period 须为 D|W|M|all，收到 ${args.period}` }));
    return 1;
  }

  const result = {
    ts_code: meta.ts_code || meta.symbol || null,
    adj: meta.adj || "backward",
    bands: {},
  };

  for (const p of periods) {
    const defaults = DEFAULTS[p];
    const window = args.window != null ? Number(args.window) : defaults.window;
    const nbdev = args.nbdev != null ? Number(args.nbdev) : defaults.nbdev;
    const series = resampleCloses(bars, p);
    result.bands[p] = { period: p, ...bollinger(series, window, nbdev) };
  }

  console.log(JSON.stringify(result, null, 2));
  return 0;
}

process.exit(main());
